### DISCUSSION TITLE
What additional request attributes should we capture?

### DISCUSSION THREAD

**Discussion by johnspurlock-skymethod** [skymethod/op3#7] [General]
Right now, OP3 follows the principle of _Data minimization_, and only captures/stores the small subset of the request attributes that it needs to serve back in the API, no more.

We have many other request-level attributes available to us, but now are completely ignoring/dropping them.  Some of these attributes should probably be considered safe to process as they have little to no listener privacy impact, and would make our eventual show/episode-level analytics much more useful.

I'd like to arrive at a consensus as to which of the following attributes are "no-brainers" to start storing/returning, those that are "on-the-line", could go either way, and those we should always avoid storing.  The goal is to provide maximum utility without going over the _creepy_ line wrt listeners' expectations.

Here are the attributes we are _not_ currently capturing for each redirect request, but _could_ in the future:
  - `asn`: _(number)_ the ASN of the incoming request (e.g. `395747`)
  - `asOrganization`: _(string)_ The organization which owns the ASN of the incoming request (e.g. `Google Cloud`)
  - `country`:  _(string)_ two-letter country code (e.g. `US`)
  - `isEUCountry`:  _(boolean)_ Whether the country is in the EU
  - `city`:  _(string)_ (e.g. `Austin`)
  - `continent`:  _(string)_ (e.g. `NA` for North America)
  - `latitude`:  _(string)_ (e.g. `30.27130`)
  - `longitude`:  _(string)_ (e.g. `-97.74260`)
  - `postalCode`:  _(string)_ (e.g. `78701`)
  - `metroCode`:  _(string)_ Metro code (DMA) (e.g. `635`)
  - `region`:  _(string)_ If known, the ISO 3166-2 name for the first level region (e.g. `Texas`)
  - `regionCode`:  _(string)_ If known, the ISO 3166-2 code for the first-level region (e.g. `TX`)
  - `timezone`: _(string)_ (e.g. `America/Chicago`)
 

To me: `country`, `continent`, `region/regionCode`, and `timezone` seem like no-brainers. 

I think `metroCode` would also be broad enough to be non-identifying but still very useful for folks building on top of OP3.

`asn` would be potentially useful for tagging traffic coming from cloud services in the future.

What does everyone think?

**Comment by jsonperl** [comment]
timezone could be nice for backing out questions like "what time of day do folks listen to my episode" without going through region gymnastics.

**Reply by johnspurlock-skymethod** [reply]
ah that's true, we could even derive a `localRequestTime` in the data API using the associated timezone.  Hmm

**Reply by jamescridland** [reply]
Yes, I see that being really useful.

(Technically: "what time do people download my episode", which isn't the time they listen to it.)

**Comment by evoterra** [comment]
I’m not sure how accurate lon/lat and other values are anyhow. Cell traffic, corporate routers… so I’d be shy of data that’s not a certainty.

**Reply by johnspurlock-skymethod** [reply]
Thanks Evo, makes sense to keep the list of captured attributes small and stable.  Garbage in, garbage out

**Comment by tomrossi7** [comment]
I don't have any hesitations about privacy with that dataset. Some of it may be redundant (e.g. a country will only ever be on one continent), but maybe you just want to make it nicer for the API to do some of those lookups for them?

_I’m not sure how accurate lon/lat and other values are anyhow._ 
Geolookup varies greatly on how accurate it is. Sometimes is very granular and accurate and other times its only at the region level or very granular, but a totally different city 🤣. At the end of the day, all location data should be taken with a grain of salt.

**Reply by johnspurlock-skymethod** [reply]
Thanks - I see nothing but downsides to keeping lat/lons at this point

**Comment by jamescridland** [comment]
Country, continent, timezone, and possibly regioncode for me (I wonder whether that's just a US thing). I'd be very wary of using anything more granular: while we understand that lat/lon or name of place is only approximate, the optics of it are that "OMG we are being tracked", and I'm not sure that helps anyone.

For me, for example, I'd expect "Brisbane" would show up, but in reality, knowing the datasets, it'll be the small suburb I live in, Bardon, which has a very low population of just 9,500 people. (Let's not forget - the podcast hosting company can absolutely show that to their customers if they want, but I'm a little more wary of this being open data available everywhere. The fewer objections people have to joining OP3, the better, as far as I'm concerned.)

asn may be useful to help combat fraud, and I think it worth including that. That could directly lead back to the (erroneous, in my view) "disregard everything from Amazon AWS" rule that the IAB has.

Other suggestions: `app` and `os` would be great, though that would require matching work done with the OPAWG list or similar, and would, I think, be out of scope.

**Reply by johnspurlock-skymethod** [reply]
Regarding region code: mind hitting [`https://workers.cloudflare.com/cf.json`](https://workers.cloudflare.com/cf.json) and letting me know your `region` and `regionCode` fields?

For the US it is effectively "US State", I wonder what it is for you.

Also, I'd be curious in the `metroCode` returned, which is the Neilsen media market code I believe.  Wonder if this is large enough to be ok.

**Reply by johnspurlock-skymethod** [reply]
`app` and `os` are certainly planned as derived fields from User-Agent, which OP3 already captures, so not really net new from a privacy POV.  I'd put those in the bucket of the other derived fields like show category from the feed.

**Reply by jamescridland** [reply]
Here's data from my mobile connection. I'll post (censored) results from my fixed connection a little later.

The lat/long places me in central Brisbane (in the middle of the river).

The bot score and trust score might be useful, as might the corporate proxy data.

`
{"clientTcpRtt":1,"longitude":"153.03250","latitude":"-27.46790","tlsCipher":"AEAD-AES128-GCM-SHA256","continent":"OC","asn":133612,"clientAcceptEncoding":"gzip, deflate, br","country":"AU","tlsClientAuth":{"certIssuerDNLegacy":"","certIssuerSKI":"","certSubjectDNRFC2253":"","certSubjectDNLegacy":"","certFingerprintSHA256":"","certNotBefore":"","certSKI":"","certSerial":"","certIssuerDN":"","certVerified":"NONE","certNotAfter":"","certSubjectDN":"","certPresented":"0","certRevoked":"0","certIssuerSerial":"","certIssuerDNRFC2253":"","certFingerprintSHA1":""},"tlsExportedAuthenticator":{"clientFinished":"31f97bb5bc41a6fcec69c494f5869e248c96752af3c4bc606ddcce27d8a2100c","clientHandshake":"68d4936dda263f120413e325321221d8b511c30532ee66a1891f715c84c0843e","serverHandshake":"196576d16e6d75f25f97115a9cc9d27354dc83410331059f3c9501218ec80add","serverFinished":"261fb1ad1581780c52b588782b96af15c423285bcf69efeb6fd65d6ff6628252"},"tlsVersion":"TLSv1.3","colo":"BNE","timezone":"Australia/Brisbane","city":"Brisbane","httpProtocol":"HTTP/2","edgeRequestKeepAliveStatus":1,"requestPriority":"weight=256;exclusive=1","botManagement":{"corporateProxy":false,"staticResource":false,"verifiedBot":false,"score":93},"clientTrustScore":93,"region":"Queensland","regionCode":"QLD","asOrganization":"Vodafone Australia","postalCode":"4000"}
`

**Reply by johnspurlock-skymethod** [reply]
Interesting, thanks - `region` is Queensland, which seems both correct and suitably broad.  And no `metroCode` at all, I guess those are US only (strictly Nielsen markets)

**Reply by jamescridland** [reply]
And from my fixed connection/Chrome (this is it in full, there's nothing really to censor)

`{"clientTcpRtt":13,"longitude":"153.00420","latitude":"-27.47500","tlsCipher":"AEAD-AES128-GCM-SHA256","continent":"OC","asn":4764,"clientAcceptEncoding":"gzip, deflate, br","country":"AU","tlsClientAuth":{"certIssuerDNLegacy":"","certIssuerSKI":"","certSubjectDNRFC2253":"","certSubjectDNLegacy":"","certFingerprintSHA256":"","certNotBefore":"","certSKI":"","certSerial":"","certIssuerDN":"","certVerified":"NONE","certNotAfter":"","certSubjectDN":"","certPresented":"0","certRevoked":"0","certIssuerSerial":"","certIssuerDNRFC2253":"","certFingerprintSHA1":""},"tlsExportedAuthenticator":{"clientFinished":"ef184251b452ca0d35085abeb349dc0f443199af1fac2e6ca7b45c613e45602a","clientHandshake":"0eb8af57b4619919ee97b7b2fc3d48ab168ebb0f9fc0f6f3b02bf138923d31ea","serverHandshake":"a0731ab0ddc37bcc81ce96625614bffc77e40e908a975c2deb66d18c532cc583","serverFinished":"d6121a657abd6dcfa780b9487f421c42e0b0a944ba959f95fb23dccfe32b9e5d"},"tlsVersion":"TLSv1.3","colo":"BNE","timezone":"Australia/Brisbane","city":"Brisbane","httpProtocol":"HTTP/2","edgeRequestKeepAliveStatus":1,"requestPriority":"weight=256;exclusive=1","botManagement":{"corporateProxy":false,"staticResource":false,"verifiedBot":false,"score":91},"clientTrustScore":91,"region":"Queensland","regionCode":"QLD","asOrganization":"Aussie Broadband","postalCode":"4064"}`

The postalCode is not correct ([4064 is a neighbouring one](https://www.google.com/maps/place/Milton+QLD+4064/@-27.4631605,152.9858189,15z/data=!4m5!3m4!1s0x6b915a016c440385:0x1c02a35e27d73820!8m2!3d-27.4611923!4d153.0000416)), but that's a population of 8,500 and I would, again, suggest that this is too small to include.

The lat/lon again places me in the middle of the Brisbane River, but in Milton.

**Reply by jamescridland** [reply]
Interestingly, a "where is my IP" lookup places my IP (IPv6) in Melbourne VIC, Australia - 1,021 miles away and an 18hr 27min drive from here.

**Reply by johnspurlock-skymethod** [reply]
What do you think about `city`, then?  Too creepy?  I think US metro seems useful and broad enough, but it's US only.

City, of course, would be useful for building "hot in cleveland" type services - where a massive state like Queensland might not be as germane

**Reply by jsonperl** [reply]
Living in Cleveland, I would welcome this functionality! 🤣

**Reply by johnspurlock-skymethod** [reply]
Right on (I grew up there): too bad github does not have a "river on fire" reaction emoji

**Comment by drpancake** [comment]
I agree that storing `latitude` and `longitude` creates more of a liability (in terms of perception of the project) than it's worth. City-level granularity seems fine. The other attributes look useful and non-controversial to me.

**Comment by daveajones** [comment]
I would be careful about including `asn` as that can help an attacker derive the secret key used to hash the IP's.  I think you are using SHA-1 for the hash, so any extra bit of input info will make cracking it easier.

**Reply by daveajones** [reply]
An alternative would be to just mark the requests as being from a hyperscaler or not.  Or maybe mark the source as residential/hyperscaler/business/etc. I know some of that info is available as lists, but not sure how fine grained it is.

**Reply by johnspurlock-skymethod** [reply]
Interesting, right it's a hmac-sha1, so it would still be very difficult, but more info signal to be sure.  `asn` could be a field that we capture (and disclose), but use only for tagging, and not returning directly in any api calls.

**Comment by johnspurlock-skymethod** [comment]
Alright, just pushed a release yesterday that captures more of these non-listener-identifying attributes on every request served via OP3.

https://op3.dev/releases#2022-10-20

Thanks to everyone for chiming in!
